lib/auth/auth_with_roles.go

```diff
@@ -4607,6 +4621,48 @@ func (a *ServerWithRoles) GetApplicationServers(ctx context.Context, namespace s
 	return filtered, nil
 }
 
+// GetAppServersAndSAMLIdPServiceProviders returns a list containing all registered AppServers and SAMLIdPServiceProviders.
+func (a *ServerWithRoles) GetAppServersAndSAMLIdPServiceProviders(ctx context.Context, namespace string) ([]types.AppServerOrSAMLIdPServiceProvider, error) {
+	appservers, err := a.authServer.GetApplicationServers(ctx, namespace)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var appAndSPs []types.AppServerOrSAMLIdPServiceProvider
+	// Convert the AppServers to AppServerOrSAMLIdPServiceProviders.
+	for _, appserver := range appservers {
+		appServerV3 := appserver.(*types.AppServerV3)
+		appAndSP := &types.AppServerOrSAMLIdPServiceProviderV1{
+			AppServerOrSP: &types.AppServerOrSAMLIdPServiceProviderV1_AppServer{
+				AppServer: appServerV3,
+			},
+		}
+		appAndSPs = append(appAndSPs, appAndSP)
+	}
+
+	// Only add SAMLIdPServiceProviders to the list if the caller has an enterprise license since this is an enteprise-only feature.
+	if modules.GetModules().BuildType() == modules.BuildEnterprise {
+		// Only attempt to list SAMLIdPServiceProviders if the caller has the permission to.
+		if err := a.action(namespace, types.KindSAMLIdPServiceProvider, types.VerbList, types.VerbRead); err == nil {
```

Comment:
why does listing this require a read as well? or i guess why doesn't `ListSAMLIdPServiceProviders` check for a read as well? idk if things changed, but the only other resource I know that only checks for `list` is [nodes](https://github.com/gravitational/teleport/blob/7ccb7ef891c170d1128909fc829db352bd2b33ea/lib/auth/auth_with_roles.go#L1658), i remember people were questioning it too (hence the detailed comment)

Reply:
I was going with the same pattern as with `GetApplicationServers`. I removed the requirement for `read` so that it has the same permission requirements as `ListSAMLIdPServiceProviders` which is used in `tctl`.